src/zarr/core/type_check.py

```diff
@@ -0,0 +1,505 @@
+    _seen: set[Any] | None = None,
+) -> Any:
+    """
+    Resolve type hints and ForwardRef. Maintains a cache of resolved types to avoid infinite recursion.
+    """
+    if _seen is None:
+        _seen = set()
+
+    # Use a more robust tracking mechanism
+    type_repr = repr(tp)
+    if type_repr in _seen:
+        # Return Any for recursive types to break the cycle
+        return Any
+
+    _seen.add(type_repr)
+
+    try:
+        return _resolve_type_impl(tp, type_map, globalns, localns, _seen)
+    finally:
+        _seen.discard(type_repr)
+
+
+def _resolve_type_impl(
+    tp: Any,
+    type_map: dict[TypeVar, Any] | None,
+    globalns: dict[str, Any] | None,
+    localns: dict[str, Any] | None,
+    _seen: set[str],
+) -> Any:
+    """
+    Internal implementation of type resolution.
+    """
+    # Substitute TypeVar
+    if isinstance(tp, TypeVar):
+        resolved = type_map.get(tp, tp) if type_map else tp
+        if isinstance(resolved, TypeVar) and resolved is tp:
+            return tp  # <-- keep literal TypeVar until check
+        return _resolve_type(resolved, type_map, globalns, localns, _seen)
+
+    # Handle string-based unions safely
+    if isinstance(tp, str) and " | " in tp:
+        parts = [p.strip() for p in tp.split("|")]
+        resolved_parts = tuple(
+            _resolve_type(
+                eval(p, globalns or {}, localns or {}), type_map, globalns, localns, _seen
+            )
+            for p in parts
+        )
+        return typing.Union[resolved_parts]  # noqa: UP007
+
+    # Evaluate ForwardRef
+    if isinstance(tp, (ForwardRef, str)):
+        ref = tp if isinstance(tp, ForwardRef) else ForwardRef(tp)
+        # Use frozenset to avoid issues with mutable default arguments
+        tp = evaluate_forward_ref(ref, globals=globalns, locals=localns)
+
+    # Recurse into Literal
+    origin = get_origin(tp)
+    args = get_args(tp)
+    if origin is Literal:
+        # Pass literal arguments through as-is, they are values, not types to resolve.
+        return Literal.__getitem__(args)
+
+    # Handle types.UnionType (Python 3.10+ union syntax like str | int)
+    if isinstance(tp, types.UnionType):
+        # Don't try to reconstruct UnionType, convert to typing.Union
+        resolved_args = tuple(_resolve_type(a, type_map, globalns, localns, _seen) for a in args)
+        return typing.Union[resolved_args]  # noqa: UP007
+
+    # Recurse into other generics
+    if origin and args:
+        new_args = tuple(_resolve_type(a, type_map, globalns, localns, _seen) for a in args)
+        # Special handling for single-argument generics like NotRequired, ReadOnly
+        if len(new_args) == 1:
+            return origin[new_args[0]]  # Pass single argument, not tuple
+        else:
+            return origin[new_args]  # Pass tuple for multi-argument generics
+
+    return tp
+
+
+def check_type(obj: Any, expected_type: Any, path: str = "value") -> TypeCheckResult:
```

Comment:
Can we narrow the type on `expected_type` here? And do we gain anything if we do?

I'd like to minimize complexity. We know we have a relatively constrained set of expected objects we're going to be passing here (node metadata, codec configuration, ...). If there's anything we can leverage to make this simpler, let's do it.

Reply:
i think you're kind of asking if we could define a "type-checkable type", i.e. a union of all the types we support runtime type checking for. I'd like that, but I haven't looked into setting that up yet. I don't know what the blockers would be, and I think it would be potentially helpful to more clearly define the scope of this checker.

although, in its current implementation, any type that isn't associated with a specific checking routine falls back to `isinstance`, which does work for pretty much all user-defined classes...